### ISSUE TITLE
[no-namespace] ignore ambient namespaces

### ISSUE THREAD

**Issue by mindplay-dk** [typescript-eslint/typescript-eslint#4556]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": [
    "@typescript-eslint/no-namespace"
  ]
}
```

```TS
export namespace Hello {
  export type A = "a";
  export type B = "b";
}

export class Hello {
  a: Hello.A = "a";
  b: Hello.B = "b";
}
```

**Expected Result**

This namespace is fully ambient - as seen in the [playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAOwIYFtgGcxIMbDgCWABsiI4BvAWACg45RJY4YBPMPAQTgF44AiJHwDcNOg2jxW7OACEe-AEbCaAXxo1xTbESTp0BYqQqi4SAFwGSEAHRdeA5bTgKLhK9bn2lI6mupA), this does not result in any code being emitted by TS, and therefore is not problematic - it effectively results in the same thing as `declare namespace` in `.d.ts` files.

In other words, because the namespace contains only ambient declarations, it should itself be considered ambient - it's being used only as a means of logically grouping some related types under a parent namespace.

**Actual Result**

As seen in the [playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=KYDwDg9gTgLgBAOwIYFtgGcxIMbDgCWABsiI4BvAWACg45RJY4YBPMPAQTgF44AiJHwDcNOg2jxW7OACEe-AEbCaAXxo1xTbESTp0BYqQqi4SAFwGSEAHRdeA5bTgKLhK9bn2lI6mupA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYECGAtojFUnCAFwhwopkogC+QA&tsConfig=N4XyA), this is being caught and reported like a concrete namespace object.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.12.0` |
| `@typescript-eslint/parser`        | `5.12.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.9.0` |
| `node`                             | `14.17.6` |

**Comment by bradzacher** [comment]
I'm not sure I understand why you would like to use a namespace within a file instead of using modules directly - even for types.
Specifically - I don't understand why not emitting any code makes a type-only namespace any better.

- What is the difference between using a namespace for types vs using a namespace for values?
- Why not just use standard module files and import them to create namespaces, as you would for values?

The rule doesn't exist because of any runtime code generated for namespaces. It exists because *generally* modules are just an all-round better way of doing things.

----

As a side-note your example also uses declaration merging; which is a concept I generally advise against because of the complexities it invites and the large potential for unsafety or just plain weird behaviour.

**Comment by mindplay-dk** [comment]
> I'm not sure I understand why you would like to use a namespace within a file instead of using modules directly

I wanted to keep them grouped where they get imported.

This project has an internal model and an external/exported model - for example, the internal model and public models might both have a `Product` type, but the internal model has `created: Date` and the exported model has `created: string`.

I need to work with both types from within the module that makes the conversions, and I wanted them grouped, so that function signatures read e.g. `function export(product: Product): Output.Product`.

I couldn't find any other way to do it, so I resorted to namespaces.

I didn't realize it was doing declaration merging - I thought `export namespace` wasn't producing anything, but apparently it produces `{}`, so it's not ambient at all.

Anyhow, yes, maybe I should just export the types without the wrapper namespace, and then `import * from "x" as Output`.

I'm not sure why `namespace` is considered so toxic? It's just a convenient way of declaring a const and properties with types in a single declaration, right? Except you can't put types in a const, which is why I resorted to this in the first place.

Anyhow, importing the whole module will work too, I guess. Feel free to close. 🙂

**Comment by bradzacher** [comment]
> I'm not sure why namespace is considered so toxic?

It's not toxic, it was just syntax added before ES6 modules as a way to "bundle" things in the global namespace underneath sub namespaces (which was the way the world worked at the time!).

But with the advent of modules - using namespaces to group things **generally** is a code smell because using actual modules and grouping via an import star is **generally** better design.

The semantics of it are very confusing to people because of things like declaration merging (which people don't understand well) or because you can `export let` which has different semantics to an `export let` from a module.

All that being said - it is certainly a powerful tool when used correctly. But "correctly" is generally a rare edge-case, so it's just banned in most codebases to push for better code patterns.

It's up to you whether or not you want to ban them in your codebase.

**Comment by JoshuaKGoldberg** [comment]
Closing as aged away since it's been about two months and nobody else has expressed interest. Thanks all!
